src/main/java/org/candlepin/jackson/GuestIdDeserializer.java

```diff
@@ -36,36 +35,41 @@
  */
 public class GuestIdDeserializer extends JsonDeserializer<GuestIdDTO> {
     private static Logger log = LoggerFactory.getLogger(GuestIdDeserializer.class);
-    private static ObjectMapper mapper = ObjectMapperFactory.getObjectMapper();
+
+    private ObjectReader reader;
+
+    /**
+     * Creates a new GuestIdDeserializer using the specified reader as the
+     * default/base reader for processing guest IDs received in object form.
+     */
+    public GuestIdDeserializer(ObjectReader reader) {
+        if (reader == null) {
+            throw new IllegalArgumentException("reader is null");
+        }
+
+        this.reader = reader;
+    }
 
     @Override
     public GuestIdDTO deserialize(JsonParser parser, DeserializationContext context)
         throws IOException, JacksonException {
-        TreeNode rootNode = parser.readValueAsTree();
-        if (rootNode.isValueNode()) {
-            return parseValueNode(rootNode, parser);
+        JsonNode node = context.readTree(parser);
+        if (node.isTextual()) {
+            log.debug("Processing node as a guest ID string: {}", node);
+
+            return new GuestIdDTO()
+                .guestId(node.textValue());
         }
-        else if (rootNode.isObject()) {
-            log.debug("Processing GuestIdDTO as a containing object node.");
-            return mapper.convertValue(rootNode, GuestIdDTO.class);
+        else if (node.isObject()) {
+            log.debug("Processing node as a GuestIdDTO instance: {}", node);
+
+            return reader.readValue(node, GuestIdDTO.class);
         }
         else {
```

Comment:
We can actually drop the "else" here and just throw the exception directly from this point. Saves us a bit of indentation.

Reply:
Updated